File: collect_app/src/main/java/org/odk/collect/android/widgets/BaseArbitraryFileWidget.java
package org.odk.collect.android.widgets;

import android.content.Context;

import org.javarosa.core.model.data.IAnswerData;
import org.javarosa.core.model.data.StringData;
import org.odk.collect.android.formentry.questions.QuestionDetails;
import org.odk.collect.android.utilities.QuestionMediaManager;
import org.odk.collect.android.widgets.interfaces.FileWidget;
import org.odk.collect.android.widgets.interfaces.WidgetDataReceiver;
import org.odk.collect.android.widgets.utilities.WaitingForDataRegistry;

import java.io.File;

import timber.log.Timber;

public abstract class BaseArbitraryFileWidget extends QuestionWidget implements FileWidget, WidgetDataReceiver  {
    private final QuestionMediaManager questionMediaManager;
    protected final WaitingForDataRegistry waitingForDataRegistry;

    protected File answerFile;

    public BaseArbitraryFileWidget(Context context, QuestionDetails questionDetails, QuestionMediaManager questionMediaManager,
                                   WaitingForDataRegistry waitingForDataRegistry) {
        super(context, questionDetails);
        this.questionMediaManager = questionMediaManager;
        this.waitingForDataRegistry = waitingForDataRegistry;
    }

    @Override
    public IAnswerData getAnswer() {
        return answerFile != null ? new StringData(answerFile.getName()) : null;
    }

    @Override
    public void deleteFile() {
        questionMediaManager.deleteAnswerFile(getFormEntryPrompt().getIndex().toString(), answerFile.getAbsolutePath());
        answerFile = null;
        hideAnswerText();
    }

    @Override
    public void setData(Object object) {
        if (answerFile != null) {
            deleteFile();
        }

        if (object instanceof File) {
            answerFile = (File) object;
            if (answerFile.exists()) {
                questionMediaManager.replaceAnswerFile(getFormEntryPrompt().getIndex().toString(), answerFile.getAbsolutePath());
                showAnswerText();
                widgetValueChanged();
            } else {
                answerFile = null;
                Timber.e("Inserting Arbitrary file FAILED");
            }
        } else if (object != null) {
            Timber.e("FileWidget's setBinaryData must receive a File but received: %s", object.getClass());
        }
    }

    protected void setupAnswerFile(String fileName) {
        if (fileName != null && !fileName.isEmpty()) {
            answerFile = new File(getInstanceFolder() + File.separator + fileName);
        }
    }

    protected abstract void showAnswerText();

    protected abstract void hideAnswerText();
}

File: collect_app/src/main/java/org/odk/collect/android/widgets/BaseArbitraryFileWidget.kt
package org.odk.collect.android.widgets

import android.content.Context
import org.javarosa.core.model.data.IAnswerData
import org.javarosa.core.model.data.StringData
import org.odk.collect.android.formentry.questions.QuestionDetails
import org.odk.collect.android.utilities.QuestionMediaManager
import org.odk.collect.android.widgets.interfaces.FileWidget
import org.odk.collect.android.widgets.interfaces.WidgetDataReceiver
import org.odk.collect.android.widgets.utilities.WaitingForDataRegistry
import timber.log.Timber
import java.io.File

abstract class BaseArbitraryFileWidget(
    context: Context,
    questionDetails: QuestionDetails,
    private val questionMediaManager: QuestionMediaManager,
    protected val waitingForDataRegistry: WaitingForDataRegistry,
    dependencies: Dependencies
) : QuestionWidget(context, dependencies, questionDetails), FileWidget, WidgetDataReceiver {
    var answerFile: File? = null

    override fun getAnswer(): IAnswerData? {
        return if (answerFile != null) StringData(answerFile!!.name) else null
    }

    override fun deleteFile() {
        questionMediaManager.deleteAnswerFile(
            formEntryPrompt.index.toString(),
            answerFile!!.absolutePath
        )
        answerFile = null
        hideAnswerText()
    }

    override fun setData(answer: Any?) {
        if (answerFile != null) {
            deleteFile()
        }

        if (answer is File) {
            answerFile = answer
            if (answerFile!!.exists()) {
                questionMediaManager.replaceAnswerFile(
                    formEntryPrompt.index.toString(),
                    answerFile!!.absolutePath
                )
                showAnswerText()
                widgetValueChanged()
            } else {
                answerFile = null
                Timber.e(Error("Inserting Arbitrary file FAILED"))
            }
        } else if (answer != null) {
            Timber.e(Error("FileWidget's setBinaryData must receive a File but received: " + answer.javaClass))
        }
    }

    protected fun setupAnswerFile(fileName: String?) {
        if (!fileName.isNullOrEmpty()) {
            answerFile = questionMediaManager.getAnswerFile(fileName)
        }
    }

    protected abstract fun showAnswerText()

    protected abstract fun hideAnswerText()
}
